src/libos/src/fs/inode_file.rs

```diff
@@ -113,6 +113,46 @@ impl File for INodeFile {
         Ok(*offset as i64)
     }
 
+    fn preadv(&self, bufs: &mut [&mut [u8]], offset: usize) -> Result<usize> {
+        if !self.access_mode.readable() {
+            return_errno!(EBADF, "File not readable");
+        }
+        self.offset.lock().unwrap();
+        let mut offset = offset;
+        let mut total_len = 0;
+        for buf in bufs {
+            match self.inode.read_at(offset, buf) {
+                Ok(len) => {
+                    total_len += len;
+                    offset += len;
+                }
+                Err(_) if total_len != 0 => break,
+                Err(e) => return Err(e.into()),
+            }
+        }
+        Ok(total_len)
+    }
+
+    fn pwritev(&self, bufs: &[&[u8]], offset: usize) -> Result<usize> {
+        if !self.access_mode.writable() {
+            return_errno!(EBADF, "File not writable");
+        }
+        self.offset.lock().unwrap();
```

Comment:
The lock here seems to be useless, because the pwritev does not change the seek offset of the opened file.